File: Sources/CountriesFlagsObjC/source/Countries/UIImage+DemocraticRepublicCongo.m
//
//  UIImage+DemocraticRepublicCongo.m
//  CountriesFlags
//

#import "Countries/UIImage+DemocraticRepublicCongo.h"

#import "Extensions/UIBezierPath+Star.h"

@implementation UIImage (DemocraticRepublicCongo)

+ (UIImage *)flagDemocraticRepublicCongoWithSize:(CGSize)size {
    UIColor* colorBlue = [UIColor colorWithRed:0.0f
                                         green:127.0f/255.0f
                                          blue:1.0f
                                         alpha:1.0f];
    UIColor* colorRed = [UIColor colorWithRed:206.0f/255.0f
                                        green:16.0f/255.0f
                                         blue:33.0f/255.0f
                                        alpha:1.0f];
    UIColor* colorYellow = [UIColor colorWithRed:247.0f/255.0f
                                         green:214.0f/255.0f
                                          blue:24.0f/255.0f
                                         alpha:1.0f];
    
    UIGraphicsBeginImageContextWithOptions(size, YES, 0.0f);
    
    [colorBlue setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height));
    
    [colorYellow setFill];
    UIBezierPath* path = [UIBezierPath bezierPath];
    [path moveToPoint:CGPointMake(0.0f, size.height)];
    [path addLineToPoint:CGPointMake(0.0f, size.height / 24.0f * 18.0f)];
    [path addLineToPoint:CGPointMake(size.width / 33.0f * 31.0f, 0.0f)];
    [path addLineToPoint:CGPointMake(size.width, 0.0f)];
    [path addLineToPoint:CGPointMake(size.width, size.height / 24.0f * 6.0f)];
    [path addLineToPoint:CGPointMake(size.width / 33.0f * 2.0f, size.height)];
    [path closePath];
    [path fill];

    [colorRed setFill];
    [path removeAllPoints];
    [path moveToPoint:CGPointMake(0.0f, size.height)];
    [path addLineToPoint:CGPointMake(0.0f, size.height / 24.0f * 19.2f)];
    [path addLineToPoint:CGPointMake(size.width, 0.0f)];
    [path addLineToPoint:(CGPointMake(size.width, size.height / 24.0f * 4.8f))];
    [path closePath];
    [path fill];
    
    [colorYellow setFill];
    [[UIBezierPath starAtPoint:CGPointMake(size.height / 24.0f * 5.85f, size.height / 24.0f * 6.25f)
                        radius:size.height / 24.0f * 4.8f
                          type:FlagStarTypeFivePointed] fill];
    
    UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image != nil ? image : [UIImage new];
}

@end

File: Sources/CountriesFlags/Countries/UIImage+DemocraticRepublicCongo.swift
//
//  UIImage+DemocraticRepublicCongo.swift
//  CountriesFlags
//

#if canImport(UIKit)

import UIKit

extension UIImage {
    /**
     [wikipedia]: https://en.wikipedia.org/wiki/Flag_of_the_Democratic_Republic_of_the_Congo "wikipedia"
     
     Generates the flag of the Flag of the Democratic Republic of the Congo.
     For more information see [wikipedia].
     
     - Parameters:
        - size: image size.
     
     - Returns: Returns the flag of the Flag of the Democratic Republic of the Congo.
     */
    internal static func flagDemocraticRepublicCongo(with size: CGSize) -> UIImage? {
        let colorBlue = UIColor(red: 0.0,
                                green: 127.0 / 255.0,
                                blue: 1.0,
                                alpha: 1.0)
        let colorRed = UIColor(red: 206.0 / 255.0,
                               green: 16.0 / 255.0,
                               blue: 33.0 / 255.0,
                               alpha: 1.0)
        let colorYellow = UIColor(red: 247.0 / 255.0,
                                  green: 214.0 / 255.0,
                                  blue: 24.0 / 255.0,
                                  alpha: 1.0)

        UIGraphicsBeginImageContextWithOptions(size, false, 0.0)

        colorBlue.setFill()
        UIRectFill(CGRect(origin: .zero, size: size))

        colorYellow.setFill()
        let path = UIBezierPath()
        path.move(to: CGPoint(x: 0.0, y: size.height))
        path.addLine(to: CGPoint(x: 0.0, y: size.height / 24.0 * 18.0))
        path.addLine(to: CGPoint(x: size.width / 33.0 * 31.0, y: 0.0))
        path.addLine(to: CGPoint(x: size.width, y: 0.0))
        path.addLine(to: CGPoint(x: size.width, y: size.height / 24.0 * 6.0))
        path.addLine(to: CGPoint(x: size.width / 33.0 * 2.0, y: size.height))
        path.close()
        path.fill()

        colorRed.setFill()
        path.removeAllPoints()
        path.move(to: CGPoint(x: 0.0, y: size.height))
        path.addLine(to: CGPoint(x: 0.0, y: size.height / 24.0 * 19.2))
        path.addLine(to: CGPoint(x: size.width, y: 0.0))
        path.addLine(to: CGPoint(x: size.width, y: size.height / 24.0 * 4.8))
        path.close()
        path.fill()

        colorYellow.setFill()
        UIBezierPath.star(at: CGPoint(x: size.height / 24.0 * 5.85, y: size.height / 24.0 * 6.25),
                          radius: size.height / 24.0 * 4.8).fill()

        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return image
    }
}

#endif
